packages/core/src/useSyncState.tsx

```diff
@@ -1,74 +1,46 @@
 import * as React from 'react';
 
-const UNINTIALIZED_STATE = {};
+const createStore = <T,>(getInitialState: () => T) => {
+  const listeners: (() => void)[] = [];
 
-/**
- * This is definitely not compatible with concurrent mode, but we don't have a solution for sync state yet.
- */
-export function useSyncState<T>(initialState?: (() => T) | T) {
-  const stateRef = React.useRef<T>(UNINTIALIZED_STATE as any);
-  const isSchedulingRef = React.useRef(false);
-  const isMountedRef = React.useRef(true);
+  let state: T = getInitialState();
 
-  React.useEffect(() => {
-    isMountedRef.current = true;
+  const getState = () => state;
 
-    return () => {
-      isMountedRef.current = false;
-    };
-  }, []);
-
-  if (stateRef.current === UNINTIALIZED_STATE) {
-    stateRef.current =
-      // @ts-expect-error: initialState is a function, but TypeScript doesn't think so
-      typeof initialState === 'function' ? initialState() : initialState;
-  }
-
-  const [trackingState, setTrackingState] = React.useState(stateRef.current);
-
-  const getState = React.useCallback(() => stateRef.current, []);
+  const setState = (newState: T) => {
+    state = newState;
+    listeners.forEach((listener) => listener());
+  };
 
-  const setState = React.useCallback((state: T) => {
-    if (state === stateRef.current || !isMountedRef.current) {
-      return;
-    }
+  const subscribe = (callback: () => void) => {
+    listeners.push(callback);
 
-    stateRef.current = state;
-
-    if (!isSchedulingRef.current) {
-      setTrackingState(state);
-    }
-  }, []);
-
-  const scheduleUpdate = React.useCallback((callback: () => void) => {
-    isSchedulingRef.current = true;
-
-    try {
-      callback();
-    } finally {
-      isSchedulingRef.current = false;
-    }
-  }, []);
+    return () => {
+      const index = listeners.indexOf(callback);
 
-  const flushUpdates = React.useCallback(() => {
-    if (!isMountedRef.current) {
-      return;
-    }
+      if (index > -1) {
+        listeners.splice(index, 1);
+      }
+    };
+  };
 
-    // Make sure that the tracking state is up-to-date.
-    // We call it unconditionally, but React should skip the update if state is unchanged.
-    setTrackingState(stateRef.current);
-  }, []);
+  return {
+    getState,
+    setState,
+    subscribe,
+  };
+};
 
-  // If we're rendering and the tracking state is out of date, update it immediately
-  // This will make sure that our updates are applied as early as possible.
-  if (trackingState !== stateRef.current) {
-    setTrackingState(stateRef.current);
-  }
+export function useSyncState<T>(getInitialState: () => T) {
+  const store = React.useRef(createStore(getInitialState)).current;
 
-  const state = stateRef.current;
+  const state = React.useSyncExternalStore(
+    store.subscribe,
+    store.getState,
+    store.getState
```

Comment:
I guess we don't need the third arg if the same as the second

Reply:
there's a warning during SSR if the third argument is not provided